executor/src/witgen/affine_expression.rs

```diff
@@ -13,128 +12,116 @@ use super::{EvalError::*, EvalResult, EvalValue, IncompleteCause};
-        if self.clean {
-            return Cow::from(&self.coefficients);
+        if let Self::Constant(c) = self {
+            Some(*c)
+        } else {
+            None
         }
+    }
 
-        match &self.coefficients[..] {
-            [] => return vec![].into(),
-            [(k, v)] => return Self::clean_one(k, v).into(),
-            [(k1, v1), (k2, v2)] => return Self::clean_two((k1, v1), (k2, v2)).into(),
-            _ => {}
-        };
-
-        // Ok, this is more complicated.
-        // Remove duplicates by first sorting and then going through
-        // all adjacent pairs with equal variable IDs, adding the coefficient
-        // of the first to the second, and setting the coefficient of the
-        // first to zero.
-        // Then we filter out the zeros as a last step.
-        let mut coefficients = self.coefficients.clone();
-        coefficients.sort_unstable_by(|(k1, _), (k2, _)| k1.cmp(k2));
-        for i in 1..coefficients.len() {
-            let (first, second) = coefficients.split_at_mut(i);
-            let (k1, v1) = first.last_mut().unwrap();
-            let (k2, v2) = second.first_mut().unwrap();
-            if k1 == k2 {
-                *v2 += *v1;
-                *v1 = 0.into();
-            }
+    #[inline]
+    pub fn is_zero(&self) -> bool {
+        if let Self::Constant(n) = self {
+            n.is_zero()
+        } else {
+            false
         }
-        coefficients
-            .into_iter()
-            .filter(|(_, c)| !c.is_zero())
-            .collect()
     }
 
-    fn clean_two(a: (&K, &T), b: (&K, &T)) -> Vec<(K, T)> {
-        if a.1.is_zero() {
-            Self::clean_one(b.0, b.1)
-        } else if b.1.is_zero() {
-            Self::clean_one(a.0, a.1)
-        } else if a.0 == b.0 {
-            Self::clean_one(a.0, &(*a.1 + *b.1))
+    #[inline]
+    pub fn is_one(&self) -> bool {
+        if let Self::Constant(n) = self {
+            n.is_one()
         } else {
-            vec![(*a.0, *a.1), (*b.0, *b.1)]
+            false
         }
     }
 
-    fn clean_one(k: &K, v: &T) -> Vec<(K, T)> {
-        if v.is_zero() {
-            vec![]
-        } else {
-            vec![(*k, *v)]
+    /// check that variables are sorted, unique and non-zero (used in sanity debug asserts)
+    fn is_proper_coefficients(coefficients: &[(K, T)]) -> bool {
+        coefficients.windows(2).all(|w| w[0].0 < w[1].0)
+            && coefficients.iter().all(|(_, c)| !c.is_zero())
+    }
+
+    /// @returns the nonzero coefficients and their variable IDs (but not the offset).
+    /// The coefficients are ordered by variable
+    pub fn nonzero_coefficients(&self) -> impl Iterator<Item = &(K, T)> {
+        match self {
+            AffineExpression::Many {
+                coefficients,
+                offset: _,
+            } => {
+                debug_assert!(Self::is_proper_coefficients(coefficients));
+                Either::Right(Either::Left(coefficients.iter()))
```

Comment:
Hm, maybe I'm not too familiar with `either`, but what does this do? Is the idea to unify the returned type? Is there not a more obvious way?

Reply:
Idea is: I want to return an `impl Trait`, but there are 3 concrete implementations being returned here.
General solution in these situations is to return Box<dyn Trait>, but that is one allocation and uses dynamic dispatch.
`Either` implements `Iterator<Item=T>` when both sides implement `Iterator<T>` type.
And this works recursively for each side.
This way we can return a concrete `impl Trait`, avoiding allocations or dynamic dispatch!

Reply:
Can you add a short comment?